R/Chrome.R

```diff
@@ -2,6 +2,141 @@
+#' `...`.
+#' @param timeouts A vector of timeouts applied to each asynchronous function.
+#'     Repeated.
+#' @param cleaning_timeout The delay for cleaning Chrome.
+#' @param async Is the result a promise? Required for using `chrome_execute()`
+#'     in Shiny.
+#' @param bin Character scalar, the path to Chromium or Chrome executable.
+#' @param debug_port Integer scalar, the Chromium/Chrome remote debugging port.
+#' @param local Logical scalar, indicating whether the local version of the
+#'     protocol (embedded in `crrri`) must be used or the protocol must be
+#'     fetched _remotely_.
+#' @param extra_args Character vector, extra command line arguments passed to
+#'     Chromium/Chrome.
+#' @param headless Logical scalar, indicating whether Chromium/Chrome is launched
+#'     in headless mode.
+#' @param retry_delay Number, delay in seconds between two successive tries to
+#'     connect to headless Chromium/Chrome.
+#' @param max_attempts Logical scalar, number of tries to connect to headless
+#'     Chromium/Chrome.
+#'
+#' @return An invisible list with the values of the fulfilled promises for each
+#'     async function.d If there is only async function, the return value is the value of the
+#'     fulfilled promise.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' async_save_as_pdf <- function(url) {
+#'   function(client) {
+#'     Page <- client$Page
+#'
+#'     Page$enable() %...>% {
+#'       Page$navigate(url = url)
+#'       Page$loadEventFired()
+#'     } %...>% {
+#'       Page$printToPDF()
+#'     } %...>% {
+#'       .$data %>%
+#'         jsonlite::base64_dec() %>%
+#'         writeBin(paste0(httr::parse_url(url)$hostname, ".pdf"))
+#'     }
+#'   }
+#' }
+#'
+#' save_as_pdf <- function(...) {
+#'   list(...) %>%
+#'     purrr::map(async_save_as_pdf) %>%
+#'     chrome_execute(.list = .)
+#' }
+#'
+#' save_as_pdf("https://www.r-project.org/", "https://rstudio.com/")
+#' }
+chrome_execute <- function(
+  ..., .list = NULL, timeouts = 30, cleaning_timeout = 30, async = FALSE,
+  bin = Sys.getenv("HEADLESS_CHROME"), debug_port = 9222L, local = FALSE,
+  extra_args = NULL, headless = TRUE, retry_delay = 0.2, max_attempts = 15L
+) {
+  if(is.null(.list)) {
+    funs <- list(...)
+  } else {
+    stopifnot(is.list(.list))
+    funs <- .list
+  }
+
+  # check arguments
+  purrr::walk(funs, check_is_single_param_fun)
+  stopifnot(is.numeric(timeouts))
+  assert_that(is.number(cleaning_timeout))
+
+  # initialize objects
+  timeouts <- rep_len(timeouts, length(funs))
+  total_timeout <- sum(timeouts) + cleaning_timeout
+  results <- vector("list", length(funs))
+  curr_env <- rlang::current_env()
+  execute_fun <- function(client, index, env) {
+    fun <- purrr::pluck(funs, index)
+    delay <- purrr::pluck(timeouts, index)
+    result_saved <- promises::then(fun(client), onFulfilled = function(value) {
+      if(!is.null(value)) {
+        results <- rlang::env_get(env, "results")
```

Comment:
I think it is safer to use name parameter
```suggestion
        results <- rlang::env_get(env = env, nm = "results")
```